### PR TITLE
Create an API endpoint for checking out

### PR DESCRIPTION
**Pull request by thesuss**
PT Story: 
https://www.pivotaltracker.com/story/show/132316359

Changes proposed in this pull request:
- request spec for return from API upon checkout

What I have learned working on this feature: 
- Request specs gets easier the more of them you create

Ready for review!

### CODE REVIEW

File: spec/requests/api/v1/checkout_spec.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  let(:dish2) { create(:dish, restaurant: restaurant, name: 'Chocolate') }
+  let(:cart) { create(:shopping_cart)}
+
+  it 'returns list of items in order with price & total price' do
+    cart.add(dish, dish.price)
+    cart.add(dish2, dish2.price)
+    post '/api/v1/checkout', {params: {id: cart.id}}
+    expect(response_json).to eq({'cart_id'=>cart.id, 'dishes'=>[{'name'=>dish.name, 'price' => dish.price},
+                                                                {'name'=>dish2.name, 'price' => dish2.price}],
+                                 'total'=>2})
```
Comment (tochman) [inline spec/requests/api/v1/checkout_spec.rb]: How about we return some sort success message? And how about the `sad-path`?

File: spec/requests/api/v1/checkout_spec.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+RSpec.describe Api::V1::CartsController do
+  let(:restaurant) { create(:restaurant) }
+  let(:dish) { create(:dish, restaurant: restaurant, name: 'Spaghetti') }
```
Comment (tochman) [inline spec/requests/api/v1/checkout_spec.rb]: Why do we care about the name of the dish?

File: app/controllers/api/v1/carts_controller.rb [hunk 3]
```diff
@@ -10,11 +10,21 @@ def create
...
     dish = Dish.find(params[:dish_id])
     @cart.add(dish, dish.price)
   rescue
     render json: ({'error' => 'Dish ID was invalid'})
   end
 
+  def checkout
+    @cart = ShoppingCart.find_by(id: params[:id])
+    @cart.paid = true
+  end
```
Comment (tochman) [inline app/controllers/api/v1/carts_controller.rb:25]: Shouldn't we assign a shopping cart to a user on checkout?

Comment (tochman) [reply app/controllers/api/v1/carts_controller.rb:25]: Any action on this?

File: spec/requests/api/v1/checkout_spec.rb (hunk 2 above)
Comment (tochman) [reply spec/requests/api/v1/checkout_spec.rb]: Any action on this?

File: spec/requests/api/v1/checkout_spec.rb (hunk 1 above)
Comment (tochman) [reply spec/requests/api/v1/checkout_spec.rb]: Any action on this?

**Review by tochman — CHANGES_REQUESTED** [review]

**Review by tochman — APPROVED** [review]

File: app/controllers/api/v1/carts_controller.rb [hunk 4]
```diff
@@ -10,11 +10,37 @@ def create
...
     dish = Dish.find(params[:dish_id])
     @cart.add(dish, dish.price)
   rescue
     render json: ({'error' => 'Dish ID was invalid'})
   end
 
+  def checkout
+    @cart = ShoppingCart.find(params[:id])
+    @cart.user = User.find(params[:user_id])
+    no_dishes_error
```
Comment (tochman) [inline app/controllers/api/v1/carts_controller.rb]: I wonder if you should condition this somehow? Like:

```
@cart.paid = true unless no_dishes_error
```

**Comment by tochman** [comment]
Looks good. I have no real objections apart from the comment I made.

File: spec/requests/api/v1/checkout_spec.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    cart.add(dish, dish.price)
+    post '/api/v1/checkout', {params: {id: cart.id, user_id: 'nonsense'}}
+    expect(response_json).to eq({ 'message' => 'Cart not associated with a customer' })
+    expect(cart.paid).to eq false
+  end
+
+  it 'returns error message if there are no dishes in the cart' do
+    post '/api/v1/checkout', {params: {id: cart.id, user_id: customer.id}}
+    expect(response_json).to eq({ 'message' => 'No dishes in cart' })
+    expect(cart.paid).to eq false
```
Comment (AmberWilkie) [inline spec/requests/api/v1/checkout_spec.rb:33]: So I tested the `@cart.paid == true`... if it goes into one of the error scenarios, it doesn't set the cart as paid.

Comment (tochman) [reply spec/requests/api/v1/checkout_spec.rb:33]: I understand that but it looks a bit weird in the controller method. This is not a blocker. I just note that reading this feels weird.

**Comment by AmberWilkie** [comment]
Ok I changed that line as per your suggestion. Tests are fine.

**Pull request merged**